### PR TITLE
[New Rule] Possible FIN7 Command and Control Behavior

### LINKED ISSUE #27: [New Rule] Possible FIN7 Command and Control Behavior
## Description
This rule detects a known command and control pattern in network events. The FIN7 threat group is known to use this
command and control technique, while maintaining persistence in their target's network.

## Required Info

- **Eventing Sources:**
Packetbeat

- **Target Operating Systems:**
NA

- **Platforms**
NA

- **Target ECS Version:** 1.5.0
- **New fields required in ECS for this? Associated issue/PR:** NA

## Optional Info
- **References:**
https://www.fireeye.com/blog/threat-research/2018/08/fin7-pursuing-an-enigmatic-and-evasive-global-criminal-operation.html

## Example Data
image
image
HTTP
```
{
  "_index": "packetbeat-7.8.0-2020.06.30-000001",
  "_type": "_doc",
  "_id": "18UOJXMBrrfU-yfrPe_i",
  "_version": 1,
  "_score": null,
  "_source": {
    "@timestamp": "2020-07-06T16:55:44.190Z",
    "http": {
      "request": {
        "headers": {
          "content-length": 0
        },
        "method": "get",
        "bytes": 72
      },
      "response": {
        "body": {
          "bytes": 166
        },
        "headers": {
          "content-length": 166,
          "content-type": "text/html"
        },
        "status_phrase": "moved permanently",
        "status_code": 301,
        "bytes": 325
      },
      "version": "1.1"
    },
    "event": {
      "start": "2020-07-06T16:55:44.190Z",
      "end": "2020-07-06T16:55:44.311Z",
      "kind": "event",
      "category": "network_traffic",
      "dataset": "http",
      "duration": 120605000
    },
    "user_agent": {
      "original": "curl/7.64.1"
    },
    "url": {
      "domain": "abcd.top",
      "path": "/",
      "full": "http://abcd.top/",
      "scheme": "http"
    },
    "agent": {
      "type": "packetbeat",
      "version": "7.8.0",
      "hostname": "[redacted].lan",
      "ephemeral_id": "692f5e75-ed6a-4831-ae2c-cbbea587fc49",
      "id": "ab255210-3866-4b07-9579-91f3a78eb996",
      "name": "[redacted].lan"
    },
    "ecs": {
      "version": "1.5.0"
    },
    "destination": {
      "ip": "[redacted]",
      "port": 80,
      "domain": "abcd.top",
      "bytes": 325
    },
    "network": {
      "type": "ipv4",
      "transport": "tcp",
      "protocol": "http",
      "community_id": "1:TrKO4OuPg+btSLBLdaHudUsSBE0=",
      "bytes": 397
    },
    "source": {
      "ip": "192.168.44.144",
      "port": 51788,
      "bytes": 72
    },
    "query": "GET /",
    "method": "get",
    "type": "http",
    "client": {
      "ip": "192.168.44.144",
      "port": 51788,
      "bytes": 72
    },
    "status": "OK",
    "server": {
      "domain": "abcd.top",
      "bytes": 325,
      "ip": "[redacted]",
      "port": 80
    },
    "host": {
      "ip": [
        "fe80::aede:48ff:fe00:1122",
        "192.168.44.144",
        "fe80::80f9:15ff:fea3:de16",
        "fe80::80f9:15ff:fea3:de16",
        "fe80::ee35:843a:f08b:405d",
        "fe80::ae1:df41:6bff:4a07",
        "10.156.0.70"
      ],
      "mac": [
        "ac:de:48:00:11:22",
        "a6:83:e7:ac:3c:98",
        "a4:83:e7:ac:3c:98",
        "82:ae:16:c5:30:00",
        "82:ae:16:c5:30:01",
        "82:ae:16:c5:30:05",
        "82:ae:16:c5:30:04",
        "82:ae:16:c5:30:01",
        "06:83:e7:ac:3c:98",
        "82:f9:15:a3:de:16",
        "82:f9:15:a3:de:16"
      ],
      "hostname": "[redacted].local",
      "name": "[redacted].lan",
      "architecture": "x86_64",
      "os": {
        "platform": "darwin",
        "version": "10.15.5",
        "family": "darwin",
        "name": "Mac OS X",
        "kernel": "19.5.0",
        "build": "19F101"
      },
      "id": "7A1239AF-8E8B-5928-A112-24FDC1718CF9"
    }
  },
  "fields": {
    "tls.server_certificate.not_before": [],
    "tls.detailed.client_certificate.not_after": [],
    "file.created": [],
    "process.parent.start": [],
    "tls.server_certificate.not_after": [],
    "event.end": [
      "2020-07-06T16:55:44.311Z"
    ],
    "file.ctime": [],
    "tls.client_certificate.not_after": [],
    "tls.client.not_after": [],
    "event.ingested": [],
    "@timestamp": [
      "2020-07-06T16:55:44.190Z"
    ],
    "tls.client_certificate.not_before": [],
    "file.mtime": [],
    "event.created": [],
    "file.accessed": [],
    "tls.server.not_before": [],
    "package.installed": [],
    "tls.client.not_before": [],
    "tls.detailed.server_certificate.not_after": [],
    "event.start": [
      "2020-07-06T16:55:44.190Z"
    ],
    "tls.server.not_after": [],
    "tls.detailed.server_certificate.not_before": [],
    "process.start": [],
    "tls.detailed.client_certificate.not_before": []
  },
  "highlight": {
    "event.category": [
      "@kibana-highlighted-field@network_traffic@/kibana-highlighted-field@"
    ],
    "destination.domain": [
      "@kibana-highlighted-field@abcd.top@/kibana-highlighted-field@"
    ],
    "type": [
      "@kibana-highlighted-field@http@/kibana-highlighted-field@"
    ],
    "network.transport": [
      "@kibana-highlighted-field@tcp@/kibana-highlighted-field@"
    ]
  },
  "sort": [
    1594054544190
  ]
}
```
TLS
```
{
  "_index": "packetbeat-7.8.0-2020.06.30-000001",
  "_type": "_doc",
  "_id": "3cUOJXMBrrfU-yfrPe_i",
  "_version": 1,
  "_score": null,
  "_source": {
    "@timestamp": "2020-07-06T16:55:46.219Z",
    "type": "tls",
    "tls": {
      "resumed": false,
      "cipher": "TLS_ECDHE_RSA_WITH_AES_256_GCM_SHA384",
      "next_protocol": "h2",
      "client": {
        "ja3": "3faa4ad39f690c4ef1c3160caa375465",
        "server_name": "abcd.us",
        "supported_ciphers": [
          "TLS_ECDHE_RSA_WITH_AES_256_GCM_SHA384",
          "TLS_ECDHE_ECDSA_WITH_AES_256_GCM_SHA384",
          "TLS_ECDHE_RSA_WITH_AES_256_CBC_SHA384",
          "TLS_ECDHE_ECDSA_WITH_AES_256_CBC_SHA384",
          "TLS_ECDHE_RSA_WITH_AES_256_CBC_SHA",
          "TLS_ECDHE_ECDSA_WITH_AES_256_CBC_SHA",
          "TLS_DHE_RSA_WITH_AES_256_GCM_SHA384",
          "TLS_DHE_RSA_WITH_AES_256_CBC_SHA256",
          "TLS_DHE_RSA_WITH_AES_256_CBC_SHA",
          "TLS_ECDHE_ECDSA_WITH_CHACHA20_POLY1305_SHA256",
          "TLS_ECDHE_RSA_WITH_CHACHA20_POLY1305_SHA256",
          "TLS_DHE_RSA_WITH_CHACHA20_POLY1305_SHA256",
          "(unknown:0xff85)",
          "TLS_DHE_RSA_WITH_CAMELLIA_256_CBC_SHA256",
          "TLS_DHE_RSA_WITH_CAMELLIA_256_CBC_SHA",
          "TLS_GOSTR341001_WITH_28147_CNT_IMIT",
          "TLS_RSA_WITH_AES_256_GCM_SHA384",
          "TLS_RSA_WITH_AES_256_CBC_SHA256",
          "TLS_RSA_WITH_AES_256_CBC_SHA",
          "TLS_RSA_WITH_CAMELLIA_256_CBC_SHA256",
          "TLS_RSA_WITH_CAMELLIA_256_CBC_SHA",
          "TLS_ECDHE_RSA_WITH_AES_128_GCM_SHA256",
          "TLS_ECDHE_ECDSA_WITH_AES_128_GCM_SHA256",
          "TLS_ECDHE_RSA_WITH_AES_128_CBC_SHA256",
          "TLS_ECDHE_ECDSA_WITH_AES_128_CBC_SHA256",
          "TLS_ECDHE_RSA_WITH_AES_128_CBC_SHA",
          "TLS_ECDHE_ECDSA_WITH_AES_128_CBC_SHA",
          "TLS_DHE_RSA_WITH_AES_128_GCM_SHA256",
          "TLS_DHE_RSA_WITH_AES_128_CBC_SHA256",
          "TLS_DHE_RSA_WITH_AES_128_CBC_SHA",
          "TLS_DHE_RSA_WITH_CAMELLIA_128_CBC_SHA256",
          "TLS_DHE_RSA_WITH_CAMELLIA_128_CBC_SHA",
          "TLS_RSA_WITH_AES_128_GCM_SHA256",
          "TLS_RSA_WITH_AES_128_CBC_SHA256",
          "TLS_RSA_WITH_AES_128_CBC_SHA",
          "TLS_RSA_WITH_CAMELLIA_128_CBC_SHA256",
          "TLS_RSA_WITH_CAMELLIA_128_CBC_SHA",
          "TLS_ECDHE_RSA_WITH_RC4_128_SHA",
          "TLS_ECDHE_ECDSA_WITH_RC4_128_SHA",
          "TLS_RSA_WITH_RC4_128_SHA",
          "TLS_RSA_WITH_RC4_128_MD5",
          "TLS_ECDHE_RSA_WITH_3DES_EDE_CBC_SHA",
          "TLS_ECDHE_ECDSA_WITH_3DES_EDE_CBC_SHA",
          "TLS_DHE_RSA_WITH_3DES_EDE_CBC_SHA",
          "TLS_RSA_WITH_3DES_EDE_CBC_SHA",
          "TLS_EMPTY_RENEGOTIATION_INFO_SCSV"
        ]
      },
      "server": {
        "subject": "CN=abcd.us",
        "issuer": "CN=Encryption Everywhere DV TLS CA - G1,OU=www.digicert.com,O=DigiCert Inc,C=US",
        "not_before": "2019-11-15T00:00:00.000Z",
        "not_after": "2020-11-15T12:00:00.000Z",
        "hash": {
          "sha1": "85483664FCE4F4863C321BD9A632108ABCA4AD4A"
        }
      },
      "detailed": {
        "version": "TLS 1.2",
        "client_hello": {
          "version": "3.3",
          "supported_compression_methods": [
            "NULL"
          ],
          "extensions": {
            "signature_algorithms": [
              "rsa_pkcs1_sha512",
              "ecdsa_secp521r1_sha512",
              "(unknown:0xefef)",
              "rsa_pkcs1_sha384",
              "ecdsa_secp384r1_sha384",
              "rsa_pkcs1_sha256",
              "ecdsa_secp256r1_sha256",
              "(unknown:0xeeee)",
              "(unknown:0xeded)",
              "(unknown:0x0301)",
              "(unknown:0x0303)",
              "rsa_pkcs1_sha1",
              "ecdsa_sha1"
            ],
            "application_layer_protocol_negotiation": [
              "h2",
              "http/1.1"
            ],
            "server_name_indication": [
              "abcd.us"
            ],
            "ec_points_formats": [
              "uncompressed"
            ],
            "supported_groups": [
              "x25519",
              "secp256r1",
              "secp384r1"
            ]
          }
        },
        "server_hello": {
          "version": "3.3",
          "selected_compression_method": "NULL",
          "extensions": {
            "application_layer_protocol_negotiation": [
              "h2"
            ],
            "_unparsed_": [
              "renegotiation_info"
            ]
          }
        },
        "server_certificate": {
          "version": 3,
          "public_key_size": 2048,
          "signature_algorithm": "SHA256-RSA",
          "serial_number": "18491524053030834926156519665066902696",
          "alternative_names": [
            "abcd.us",
            "*.abcd.us"
          ],
          "public_key_algorithm": "RSA",
          "not_after": "2020-11-15T12:00:00.000Z",
          "issuer": {
            "organization": "DigiCert Inc",
            "organizational_unit": "www.digicert.com",
            "common_name": "Encryption Everywhere DV TLS CA - G1",
            "country": "US"
          },
          "subject": {
            "common_name": "abcd.us"
          },
          "not_before": "2019-11-15T00:00:00.000Z"
        },
        "server_certificate_chain": [
          {
            "subject": {
              "country": "US",
              "organization": "DigiCert Inc",
              "organizational_unit": "www.digicert.com",
              "common_name": "Encryption Everywhere DV TLS CA - G1"
            },
            "public_key_size": 2048,
            "signature_algorithm": "SHA256-RSA",
            "public_key_algorithm": "RSA",
            "not_after": "2027-11-27T12:46:10.000Z",
            "version": 3,
            "issuer": {
              "common_name": "DigiCert Global Root CA",
              "country": "US",
              "organization": "DigiCert Inc",
              "organizational_unit": "www.digicert.com"
            },
            "not_before": "2017-11-27T12:46:10.000Z",
            "serial_number": "3290217995900168375215973871519570865"
          },
          {
            "version": 3,
            "issuer": {
              "organizational_unit": "www.digicert.com",
              "common_name": "DigiCert Global Root CA",
              "country": "US",
              "organization": "DigiCert Inc"
            },
            "not_before": "2006-11-10T00:00:00.000Z",
            "public_key_size": 2048,
            "signature_algorithm": "SHA1-RSA",
            "subject": {
              "common_name": "DigiCert Global Root CA",
              "country": "US",
              "organization": "DigiCert Inc",
              "organizational_unit": "www.digicert.com"
            },
            "serial_number": "10944719598952040374951832963794454346",
            "not_after": "2031-11-10T00:00:00.000Z",
            "public_key_algorithm": "RSA"
          }
        ],
        "client_certificate_requested": false
      },
      "version": "1.2",
      "established": true,
      "version_protocol": "tls"
    },
    "source": {
      "ip": "192.168.44.144",
      "port": 51793
    },
    "client": {
      "ip": "192.168.44.144",
      "port": 51793
    },
    "ecs": {
      "version": "1.5.0"
    },
    "host": {
      "hostname": "[redacted].local",
      "architecture": "x86_64",
      "name": "[redacted].lan",
      "os": {
        "version": "10.15.5",
        "family": "darwin",
        "name": "Mac OS X",
        "kernel": "19.5.0",
        "build": "19F101",
        "platform": "darwin"
      },
      "id": "7A1239AF-8E8B-5928-A112-24FDC1718CF9",
      "ip": [
        "fe80::aede:48ff:fe00:1122",
        "192.168.44.144",
        "fe80::80f9:15ff:fea3:de16",
        "fe80::80f9:15ff:fea3:de16",
        "fe80::ee35:843a:f08b:405d",
        "fe80::ae1:df41:6bff:4a07",
        "10.156.0.70"
      ],
      "mac": [
        "ac:de:48:00:11:22",
        "a6:83:e7:ac:3c:98",
        "a4:83:e7:ac:3c:98",
        "82:ae:16:c5:30:00",
        "82:ae:16:c5:30:01",
        "82:ae:16:c5:30:05",
        "82:ae:16:c5:30:04",
        "82:ae:16:c5:30:01",
        "06:83:e7:ac:3c:98",
        "82:f9:15:a3:de:16",
        "82:f9:15:a3:de:16"
      ]
    },
    "agent": {
      "hostname": "[redacted].lan",
      "ephemeral_id": "692f5e75-ed6a-4831-ae2c-cbbea587fc49",
      "id": "ab255210-3866-4b07-9579-91f3a78eb996",
      "name": "[redacted].lan",
      "type": "packetbeat",
      "version": "7.8.0"
    },
    "server": {
      "ip": "[redacted]",
      "port": 443,
      "domain": "abcd.us"
    },
    "network": {
      "transport": "tcp",
      "protocol": "tls",
      "community_id": "1:8yX8D4MdmNgYbbyIF4oMg0yRX6k=",
      "type": "ipv4"
    },
    "event": {
      "dataset": "tls",
      "duration": 253945000,
      "start": "2020-07-06T16:55:46.219Z",
      "end": "2020-07-06T16:55:46.473Z",
      "kind": "event",
      "category": "network_traffic"
    },
    "status": "OK",
    "destination": {
      "ip": "[redacted]",
      "port": 443,
      "domain": "abcd.us"
    }
  },
  "fields": {
    "tls.server_certificate.not_before": [
      "2019-11-15T00:00:00.000Z"
    ],
    "tls.detailed.client_certificate.not_after": [],
    "file.created": [],
    "process.parent.start": [],
    "tls.server_certificate.not_after": [
      "2020-11-15T12:00:00.000Z"
    ],
    "event.end": [
      "2020-07-06T16:55:46.473Z"
    ],
    "file.ctime": [],
    "tls.client_certificate.not_after": [],
    "tls.client.not_after": [],
    "event.ingested": [],
    "@timestamp": [
      "2020-07-06T16:55:46.219Z"
    ],
    "tls.client_certificate.not_before": [],
    "file.mtime": [],
    "event.created": [],
    "file.accessed": [],
    "tls.server.not_before": [
      "2019-11-15T00:00:00.000Z"
    ],
    "package.installed": [],
    "tls.client.not_before": [],
    "tls.detailed.server_certificate.not_after": [
      "2020-11-15T12:00:00.000Z"
    ],
    "event.start": [
      "2020-07-06T16:55:46.219Z"
    ],
    "tls.server.not_after": [
      "2020-11-15T12:00:00.000Z"
    ],
    "tls.detailed.server_certificate.not_before": [
      "2019-11-15T00:00:00.000Z"
    ],
    "process.start": [],
    "tls.detailed.client_certificate.not_before": []
  },
  "highlight": {
    "event.category": [
      "@kibana-highlighted-field@network_traffic@/kibana-highlighted-field@"
    ],
    "destination.domain": [
      "@kibana-highlighted-field@abcd.us@/kibana-highlighted-field@"
    ],
    "type": [
      "@kibana-highlighted-field@tls@/kibana-highlighted-field@"
    ],
    "network.transport": [
      "@kibana-highlighted-field@tcp@/kibana-highlighted-field@"
    ]
  },
  "sort": [
    1594054546219
  ]
}
```

### PR DESCRIPTION
**Pull request by peasead**
## Issues
Resolves #27 

## Summary
This rule detects a known command and control pattern in network events. The FIN7 threat group is known to use this
command and control technique, while maintaining persistence in their target's network.

## Contributor checklist

- Have you signed the [contributor license agreement](https://www.elastic.co/contributor-agreement)? Yes
- Have you followed the [contributor guidelines](https://github.com/elastic/detection-rules/blob/main/CONTRIBUTING.md)? Yes

### CODE REVIEW

File: rules/network/command_and_control_fin7_c2_behavior.toml
Comment (brokensound77) [inline rules/network/command_and_control_fin7_c2_behavior.toml:10]: I was tempted to say remove _This rule_ from `description` and `false_positives`, but then I realized we have 22 rules where `description` starts this way. It does seem a bit unnecessary IMO, but if we are consistent either way, I think that is best (meaning if we remove here, we do everywhere). We can discuss this before deciding.

File: rules/network/command_and_control_fin7_c2_behavior.toml
Comment (rw-access) [inline rules/network/command_and_control_fin7_c2_behavior.toml:34]: missing an AND. I don't remember whether AND or OR is implicit for lucene, but we should make it explicit
```suggestion
event.category:(network OR network_traffic) AND type:(tls OR http) AND network.transport:tcp
AND destination.domain:/[a-zA-Z]{4,5}\.(pw|us|club|info|site|top)/ AND NOT destination.domain:zoom.us
```

File: rules/network/command_and_control_fin7_c2_behavior.toml
Comment (rw-access) [inline rules/network/command_and_control_fin7_c2_behavior.toml:22]: I think we could add a little more specificity to this. If we make another FIN7 C2 rule, then we'll want to differentiate the rules. Maybe adding something about the DGA or DNS-based approach to detection in the title?

File: rules/network/command_and_control_fin7_c2_behavior.toml
Comment (brokensound77) [reply rules/network/command_and_control_fin7_c2_behavior.toml:34]: I like the suggestion to add DGA to the name for cases when this FPs - since it could make triage more obvious

File: rules/network/command_and_control_fin7_c2_behavior.toml
Comment (peasead) [reply rules/network/command_and_control_fin7_c2_behavior.toml]: Agreed, added.

File: rules/network/command_and_control_fin7_c2_behavior.toml
Comment (peasead) [reply rules/network/command_and_control_fin7_c2_behavior.toml:10]: As an author and operator, I like the BLUF consistency of "this rule does this thing" and then adds some context vs. having to read through a paragraph each to ascertain what I think it means.

If we want to remove it, that's fine, but we should replace it with something we can "standardize".

Comment (brokensound77) [reply rules/network/command_and_control_fin7_c2_behavior.toml:10]: Agreed. When triaging, conciseness goes a long way and I was simply commenting that with `This rule detects ... things` vs `Detects ... things` that it is implied the description is referring to the rule. Your call either way 👍

Comment (peasead) [reply rules/network/command_and_control_fin7_c2_behavior.toml:10]: Thanks, @brokensound77 great insight.

I think we should look at standardizing descriptions, false positives, notes, etc and update existing rules once we've landed on a solution.

File: rules/network/command_and_control_fin7_c2_behavior.toml
Comment (dcode) [reply rules/network/command_and_control_fin7_c2_behavior.toml]: FWIW, if not otherwise specified, multiple terms in a Lucene query default to `AND` (e.g. `category: network type: tls`) and multiple terms within a fielded search, default to `OR` (e.g. `type: (tls http)`). `NOT` is just a negation and follows these existing rules. To be fair of the Lucene critics out there, it's actually more complicated than that because `AND`, `OR`, and `NOT` merely express a _should_ relationship by changing the scoring of a document, whereas the `+` and `-` operator specify a _must_ relationship which must evaluate true to be returned.

:themoreyouknow:

File: rules/network/command_and_control_fin7_c2_behavior.toml
Comment (dcode) [reply rules/network/command_and_control_fin7_c2_behavior.toml:10]: Agree with this. A style guide of rule writing would be helpful. Even standardizing on the ordering of search terms would be helpful to offer a common convention to make logic easier to understand across the entire ruleset.

Comment (brokensound77) [reply rules/network/command_and_control_fin7_c2_behavior.toml:10]: We do have some verbiage on query ordering: https://github.com/elastic/detection-rules/blob/main/CONTRIBUTING.md#using-elastic-common-schema-ecs 

As well as writing style: https://github.com/elastic/detection-rules/blob/main/CONTRIBUTING.md#writing-style

But there is def room for standardization, which may evolve organically over time

File: rules/network/command_and_control_fin7_c2_behavior.toml
Comment (brokensound77) [inline rules/network/command_and_control_fin7_c2_behavior.toml:35]: any opportunities to tighten this down to more specific fin7 activity to avoid FPs like `zoom.us`?

Comment (peasead) [reply rules/network/command_and_control_fin7_c2_behavior.toml:35]: When I looked at a large dataset, the only FPs were from the `us` TLD and it was Zoom.

We could eliminate that TLD, but we could miss some C2 activity.

**Review by brokensound77 — APPROVED** [review]
Like the `note`.

LGTM 👍

**Pull request merged**
